rust/frontend/src/function.rs

```diff
@@ -0,0 +1,476 @@
+//! This module provides idiomatic Rust API for creating and working with TVM functions.
+//!
+//! For calling an already registered TVM function use [`function::Builder`] and to register
+//! a TVM packed function from Rust either use [`function::register`] or the
+//! macro [`register_global_func`].
+//!
+//! See the `tests` and `examples` repository for usage examples.
+
+use std::ffi::{CStr, CString};
+use std::mem;
+use std::os::raw::{c_char, c_int, c_void};
+use std::ptr;
+use std::slice;
+use std::str;
+use std::sync::Mutex;
+
+use ts;
+
+use ty::TypeCode;
+use value::TVMValue;
+use value::ValueKind;
+use Error;
+use Module;
+use Result;
+use TVMArgValue;
+use TVMRetValue;
+
+lazy_static! {
+    static ref GLOBAL_FUNCTION_NAMES: Mutex<Vec<&'static str>> = list_global_func_names();
+}
+
+fn list_global_func_names() -> Mutex<Vec<&'static str>> {
+    let mut out_size = 0 as c_int;
+    let mut name = ptr::null() as *const c_char;
+    let mut out_array = &mut name as *mut _;
+    check_call!(ts::TVMFuncListGlobalNames(
+        &mut out_size as *mut _,
+        &mut out_array as *mut _
+    ));
+    let list = unsafe { slice::from_raw_parts(out_array, out_size as usize) };
+    let list = list
+        .iter()
+        .map(|&p| unsafe { CStr::from_ptr(p) })
+        .map(|cs| cs.to_bytes())
+        .map(|bs| str::from_utf8(bs).unwrap())
+        .collect();
+    Mutex::new(list)
+}
+
+/// Returns a registered TVM function by name.
+pub fn get_global_func(name: &str, is_global: bool, allow_missing: bool) -> Option<Function> {
+    let name = CString::new(name).unwrap();
+    let mut handle = ptr::null_mut() as ts::TVMFunctionHandle;
+    check_call!(ts::TVMFuncGetGlobal(
+        name.as_ptr() as *const c_char,
+        &mut handle as *mut _
+    ));
+    if !(handle.is_null()) {
+        mem::forget(name);
+        return Some(Function::new(handle, is_global, false));
+    } else {
+        if allow_missing {
+            return None;
+        } else {
+            panic!("Cannot find global function {:?}", name);
```

Comment:
Good point! but I followed the [python](https://github.com/dmlc/tvm/blob/master/python/tvm/_ffi/function.py#L190) and [java](https://github.com/dmlc/tvm/blob/master/jvm/core/src/main/java/ml/dmlc/tvm/Function.java#L63) sides. Actually I'm think it might be better to return `Result<Option<Function>>` instead.

Reply:
Plain `Option` is the better approach. The reason why it's useful to have `allow_missing` in other languages is because there isn't already an explicit `Option` which needs to be unwrapped. The python and java "option" auto-unwraps.